### PR TITLE
feat: collect shared logic for blob storage and minio

### PR DESCRIPTION
**Pull request by chuang8511**
Because

- blob storage will be used in different service

This commit

- collect get and upload data function

Note
- For minio pkg, the current pkg is not fit with the current design of blob storage. We will refactor that part with run logging.
- We will need to add the test code

### CODE REVIEW

File: blobstorage/blobstorage.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	resp, err := client.Do(req)
+
+	if err != nil {
+		return fmt.Errorf("uploading blob: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		body, _ := io.ReadAll(resp.Body)
+		return fmt.Errorf("upload failed with status %d: %s", resp.StatusCode, string(body))
```
Comment (jvallesm) [inline blobstorage/blobstorage.go:40]: nit: I'd add the body (even the status code) as a separate log. You want to keep errors that come from the same path as consistent as possible. That allows you, e.g., to group errors to see how frequent an incident is.

**Review by jvallesm — COMMENTED** [review]
A little comment on this but not a blocker. Did you extract this code from any service (because another one needs to use it) or is it something several services _will_ use in the future?

File: blobstorage/blobstorage.go (hunk 1 above)
Comment (chuang8511) [reply blobstorage/blobstorage.go]: I am not sure if I got it correctly. 
Do you mean like [this](https://github.com/instill-ai/x/pull/27/commits/48880040d83296052e5fd3a91cc54aa94877b837#diff-f08f67e3874ac803e2c4620a878a9f7d79bfc0ab9e1e6b7f91b11493c8abd833R41)?

**Comment by chuang8511** [comment]
> Did you extract this code from any service (because another one needs to use it) or is it something several services will use in the future?

I think `UploadFile` will be used in `model-backend` & `pipeline-backend`. So, I extracted it.
Is this what you are asking?

File: blobstorage/blobstorage.go (hunk 1 above)
Comment (jvallesm) [reply blobstorage/blobstorage.go]: Sort of, I'd align with the logger package we use in other repos:

```go
	err := fmt.Errorf("failed to upload file")
	logger.Error("Failed to upload file to MinIO",
		zap.String("body", string(body)),
		zap.Int("status", resp.StatusCode),
		zap.Error(err),
	)
	return err
```

**Comment by jvallesm** [comment]
> > Did you extract this code from any service (because another one needs to use it) or is it something several services will use in the future?
> 
> I think `UploadFile` will be used in `model-backend` & `pipeline-backend`. So, I extracted it. Is this what you are asking?

Yeah, my question was if this had been copied from somewhere (e.g. `api-gateway` plugin) or if it was new code. Basically I wanted to know which will be the current and future clients for this code. Thanks for the answer, it's clear 👍

File: blobstorage/blobstorage.go (hunk 1 above)
Comment (chuang8511) [reply blobstorage/blobstorage.go]: [updated commit](https://github.com/instill-ai/x/pull/27/commits/342f816ae8c35cef2531f729bf66ddeb2e81c3c7).
[update pipeline-backend](https://github.com/instill-ai/pipeline-backend/pull/844).

**Review by jvallesm — APPROVED** [review]

**Review by donch1989 — APPROVED** [review]

**Pull request merged**